packages/data-broker/src/services/dhis/pullers/AnalyticsPuller.js

```diff
@@ -0,0 +1,283 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import groupBy from 'lodash.groupby';
+import keyBy from 'lodash.keyby';
+
+import {
+  buildAnalyticsFromDhisAnalytics,
+  buildAnalyticsFromDhisEventAnalytics,
+  buildAnalyticsFromEvents,
+} from '../buildAnalytics';
+
+export class AnalyticsPuller {
+  constructor(dataSourceModel, translator) {
+    this.dataSourceModel = dataSourceModel;
+    this.translator = translator;
+  }
+
+  groupDataSourcesByDhisDataType = dataSources =>
+    groupBy(
+      dataSources,
+      d => d.config?.dhisDataType || this.dataSourceModel.getDhisDataTypes().DATA_ELEMENT,
+    );
+
+  groupIndicatorDataSourcesByPeriodType = dataSources => {
+    if (
+      dataSources.some(
+        d =>
+          d.config?.dhisDataType &&
+          d.config.dhisDataType !== this.dataSourceModel.getDhisDataTypes().INDICATOR,
+      )
+    ) {
+      throw new Error('All data sources have to be DHIS indicators');
+    }
+    return groupBy(dataSources, d => d.config?.indicator?.dataPeriodType || 'NONE');
+  };
+
+  getAnalyticsQueryParameters = (dataSources, options) => {
+    const dataElementCodes = dataSources.map(({ dataElementCode }) => dataElementCode);
+    const {
+      organisationUnitCode,
+      organisationUnitCodes,
+      period,
+      startDate,
+      endDate,
+      additionalDimensions,
+    } = options;
+    return {
+      dataElementCodes,
+      outputIdScheme: 'code',
+      organisationUnitCodes: organisationUnitCode ? [organisationUnitCode] : organisationUnitCodes,
+      period,
+      startDate,
+      endDate,
+      additionalDimensions,
+    };
+  };
+
+  pullAnalyticsForDhisDataType = async (apis, dataSources, options, dhisDataType) => {
+    const pullAnalyticsForApi = this.getPullAnalyticsForApiMethod({ ...options, dhisDataType });
+    const response = {
+      results: [],
+      metadata: {
+        dataElementCodeToName: {},
+      },
+    };
+    const pullAnalyticsForOptions = async (api, dataSourceList, pullOptions) => {
+      const { results, metadata } = await pullAnalyticsForApi(api, dataSourceList, pullOptions);
+      response.results.push(...results);
+      response.metadata = {
+        dataElementCodeToName: {
+          ...(response.metadata?.dataElementCodeToName || {}),
+          ...(metadata?.dataElementCodeToName || {}),
+        },
+      };
+    };
+    const pullForApi = async api => {
+      if (dhisDataType === this.dataSourceModel.getDhisDataTypes().INDICATOR) {
+        // Multiple DHIS Indicators may have different period types,
+        // so we have to group them by their period types and fetch them separately
+        const groupedDataSourcesByPeriodType = this.groupIndicatorDataSourcesByPeriodType(
+          dataSources,
+        );
```

Comment:
This is the logic to handle fetching multiple indicators with different submitted period types